bwpy/signal.py

```diff
@@ -0,0 +1,133 @@
+from . import mea_viewer
+import re
+import abc
+import functools
+import numpy as np
+import numpy.lib.stride_tricks as np_tricks
+
+
+__all__ = []
+
+
+def _transformer_factory(cls):
+    @functools.wraps(cls)
+    def transformer_factory(slice, *args, **kwargs):
+        return slice._transform(cls(*args, **kwargs))
+
+    return transformer_factory
+
+
+class Transformer(abc.ABC):
+    def __init_subclass__(cls, operator=None, **kwargs) -> None:
+        super().__init_subclass__(**kwargs)
+        name = operator or re.sub("([a-z0-9])([A-Z])", r"\1_\2", cls.__name__).lower()
+        globals()[name] = _transformer_factory(cls)
+        __all__.append(name)
+
+    def _apply_window(self, data, bin_size):
+        rows = data.shape[1]
+        cols = data.shape[2]
+        window_shape = (bin_size, rows, cols)
+        # sliding window returns more complex shape like (num_windows, 1, 1, rows, cols, bin_size)
+        # with the reshape we get rid of the unnecessary complexity (1, 1)
+        return np_tricks.sliding_window_view(data, window_shape).reshape(
+            -1, bin_size, rows, cols
+        )
```

Comment:
Can you move all of this to a separate class `WindowedTransformer`, that way all the classes without a window don't have a possibly confusing `_apply_window` member. It even looks like this doesn't need to be a class member at all, but could be a regular function like `get_signal_window` or even just `window`. Could you also rename `bin_size` to `window_size` or `window_length`?

Reply:
I created `WindowedTransform` that inherits from `Transform` and add the `get_signal_window` method, tell me if that suits your requests because I'm not sure